### DISCUSSION TITLE
Error Occurring on Deployed Server Heroku with Gunicorn (Flask + Socket.io)

### DISCUSSION THREAD

**Discussion by GoldmanMaya** [miguelgrinberg/Flask-SocketIO#1996] [Q&A]
Description:
I am encountering an error on a deployed server on Heroku using Flask and Socket.io.
 When a client sends a request, instead of the expected response, an error is returned. 
 
![image](https://github.com/miguelgrinberg/Flask-SocketIO/assets/117084595/e783a797-0e39-4bea-85e6-13b27c9d9a5b)

here is the procfile: 

`web: NEW_RELIC_CONFIG_FILE=newrelic.ini newrelic-admin run-program bin/qgtunnel gunicorn -w 1 --threads 4 wsgi:app`

requirements.txt:

```
-i https://pypi.org/simple
async-timeout==4.0.2 ; python_full_version <= '3.11.2'
bidict==0.22.1 ; python_version >= '3.7'
blinker==1.6.2
certifi==2023.5.7 ; python_version >= '3.6'
charset-normalizer==3.1.0 ; python_full_version >= '3.7.0'
click==8.1.3
dnspython==2.3.0
elasticsearch==7.17.9
et-xmlfile==1.1.0 ; python_version >= '3.6'
eventlet==0.33.3
flask==2.3.2
flask-cors===3.0.10
flask-oidc==1.4.0
flask-socketio==5.3.4
greenlet==2.0.2 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4'
gunicorn==20.1.0
h11==0.14.0 ; python_version >= '3.7'
httplib2==0.22.0 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'
idna==3.4 ; python_version >= '3.5'
itsdangerous==2.1.2
jinja2==3.1.2
lxml==4.9.2 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4'
markupsafe===2.1.2
marshmallow==2.16.3
newrelic==8.8.0
numpy==1.25.0 ; python_version >= '3.10'
oauth2client==4.1.3
openpyxl==3.1.2
pandas==2.0.1
pip==23.1.2
pyasn1==0.5.0 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4, 3.5'
pyasn1-modules==0.3.0 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4, 3.5'
pycountry==22.3.5
pycryptodome==3.18.0
pymongo==3.13.0
pyparsing==3.1.0 ; python_version >= '3.1'
python-dateutil==2.8.2
python-dotenv==1.0.0
python-engineio==4.4.1 ; python_version >= '3.6'
python-socketio==5.8.0 ; python_version >= '3.6'
pytrends==4.9.2
pytz==2023.3
redis==4.5.5
requests==2.31.0 ; python_version >= '3.7'
rollbar==0.16.3
rsa==4.9 ; python_version >= '3.6' and python_version < '4'
setuptools==68.0.0
simple-websocket==0.10.1
six==1.16.0 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3'
slack-webhook==1.0.7
tzdata==2023.3 ; python_version >= '2'
urllib3==1.26.16 ; python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4, 3.5'
werkzeug==2.3.6
wheel==0.40.0
wsproto==1.2.0 ; python_full_version >= '3.7.0'

```

I run the socket in this way: 
<img width="720" alt="image" src="https://github.com/miguelgrinberg/Flask-SocketIO/assets/117084595/4655a6b9-b2c3-471f-91d1-b90f45ea9970">

 It's important to note that everything works fine on localhost, but the issue arises specifically on the deployed server. I have thoroughly searched for solutions and attempted various troubleshooting steps, but none have been successful thus far. Any assistance would be greatly appreciated. Thank you!

**Comment by miguelgrinberg** [comment]
Google the error. This is a Gunicorn bug, you'll find lots of references to it.

**Reply by GoldmanMaya** [reply]
Is there any work around I can take to resolve the issue? I'm currently unable to use websockets in my Heroku application, and I haven't had any success so far.

**Reply by miguelgrinberg** [reply]
This happened many years ago, I don't really remember much of the details. But as I recall this was just an error that occured after closing a WebSocket connection, it was a nuisance more than a bug, since connections happened anyway.

**Reply by GoldmanMaya** [reply]
Thanks for the quick reply!! 
Intresting, in my case it does seem like a bug since the connection seems to fail to establish, resulting in the inability to send any data through the WebSocket.
 I wonder if  there might be an issue with my WebSocket implementation in other parts of the code. However, everything seems to work fine when I run it locally -  what could be causing this discrepancy in behavior between the local environment and the deployed one?

**Reply by miguelgrinberg** [reply]
This is a Gunicorn bug. I assume you don't use Gunicorn in your local set up? Then maybe not using Gunicorn for the production version is the answer?
